### PR TITLE
Fixed command line invocation crashes (C++)

### LINKED ISSUE #4719: Slic3r : terminate called after throwing an instance of 'Slic3r::UnknownOptionException"
### Do not report Prusa3D Slic3r bugs here without confirming it is a problem on a development release of Slic3r, or your issue will be closed. [Development builds of Slic3r](https://dl.slic3r.org/dev). [Prusa3D Issue Tracker](https://github.com/Prusa3D/Slic3r/issues)

**Describe the bug**

I've built Slic3r C++ source code with cmake on Debian 9.

When i launch, i have this message :

>     terminate called after throwing an instance of 'Slic3r::UnknownOptionException.

How to fix the options correctly?

**To Reproduce**
I launch Slic3r

**3D Model and Slic3r Configuration Export**
No model or configuration

**Expected behavior**
No exception

**Screenshots**

**Desktop (please complete the following information):**
 - OS: Debian 9
 - Version : 1.3.0-1292-g66b00c90

**Additional context**

### PR DESCRIPTION
**Pull request by zdroyer**
The patches prevent from crashing the the C++ commandline version on Linux.
The tested call format:
slic3r -g --no-dont-arrange --load <config.ini>  <input_file_stl/amf/3mf> --output <output_gcode> --post-process="/bin/true" 

Fixes #4719

### CODE REVIEW

File: xs/src/libslic3r/ConfigBase.hpp [hunk 1]
```diff
@@ -1,6 +1,10 @@
 #ifndef slic3r_ConfigBase_hpp_
 #define slic3r_ConfigBase_hpp_
 
+#include <execinfo.h>
```
Comment (lordofhyphens) [inline xs/src/libslic3r/ConfigBase.hpp]: This breaks builds on Windows.

File: xs/src/libslic3r/ConfigBase.hpp [hunk 2]
```diff
@@ -31,20 +37,40 @@ class ConfigOptionException : public std::exception {
     public:
     const t_config_option_key opt_key;
     ConfigOptionException(const t_config_option_key _opt_key)
-        : opt_key(_opt_key) {};
+        : opt_key(_opt_key) {
+        /*
```
Comment (lordofhyphens) [inline xs/src/libslic3r/ConfigBase.hpp]: Leave the code in, surrounded by `#ifdef` for SLIC3R_DEBUG

File: xs/src/libslic3r/ConfigBase.hpp [hunk 3]
```diff
@@ -31,20 +37,40 @@ class ConfigOptionException : public std::exception {
...
 class InvalidOptionException : public ConfigOptionException {
     using ConfigOptionException::ConfigOptionException;
     
-    public:
-    virtual const char* what() const noexcept {
-        std::string s("Invalid value for option: ");
-        s += this->opt_key;
-        return s.c_str();
-    }
+
```
Comment (lordofhyphens) [inline xs/src/libslic3r/ConfigBase.hpp]: Why was this removed?

File: xs/src/libslic3r/PrintGCode.cpp [hunk 4]
```diff
@@ -405,6 +405,10 @@ PrintGCode::process_layer(size_t idx, const Layer* layer, const Points& copies)
         std::vector<double> mm3_per_mm;
         for (auto region_id = 0U; region_id < _print.regions.size(); ++region_id) {
             const PrintRegion* region = _print.get_region(region_id);
+            if( region_id >= layer->region_count() ){
+		std::cerr<<"Layer #"<<layer->id()<<" doesn't have region "<<region_id<<". The layer has only "<<layer->region_count()<<" regions."<<std::endl;
```
Comment (lordofhyphens) [inline xs/src/libslic3r/PrintGCode.cpp]: Use `Slic3r::Log::error()`

File: xs/src/libslic3r/SupportMaterial.cpp [hunk 5]
```diff
@@ -69,6 +69,12 @@ SupportMaterial::generate(PrintObject *object)
     map<coordf_t, Polygons> &contact = contact_overhang.first;
     map<coordf_t, Polygons> &overhang = contact_overhang.second;
 
+    //bugfix: do not try to generate overhang if there is no contact area 
+    if( contact.empty() ){
+	std::cerr<<"Empty contact_areas of SupportMaterial for object ["<<object->model_object()->name<<"]"<<std::endl;
```
Comment (lordofhyphens) [inline xs/src/libslic3r/SupportMaterial.cpp]: Use `Slic3r::Log`

File: xs/src/libslic3r/ConfigBase.hpp [hunk 6]
```diff
@@ -1,6 +1,10 @@
 #ifndef slic3r_ConfigBase_hpp_
 #define slic3r_ConfigBase_hpp_
 
+#include <execinfo.h>
+#include <stdio.h>
```
Comment (lordofhyphens) [inline xs/src/libslic3r/ConfigBase.hpp]: This duplicated `<cstdio>`. Remove.

File: xs/src/libslic3r/ConfigBase.hpp [hunk 7]
```diff
@@ -1,6 +1,10 @@
...
+#include <stdlib.h>
```
Comment (lordofhyphens) [inline xs/src/libslic3r/ConfigBase.hpp]: What is getting used here that is in stdlib? If it's actually getting used, should be `<cstdlib>`

File: xs/src/libslic3r/ConfigBase.cpp [hunk 8]
```diff
@@ -916,7 +921,7 @@ ConfigOptionPoint::deserialize(std::string str, bool append) {
         this->value.x = boost::lexical_cast<coordf_t>(tokens[0]);
         this->value.y = boost::lexical_cast<coordf_t>(tokens[1]);
     } catch (boost::bad_lexical_cast &e){
-        std::cout << "Exception caught : " << e.what() << std::endl;
+        std::cout << "Config option deserialisation error of ["<<str<<"]  : " << e.what()<<" (expected 2D point)" << std::endl;
```
Comment (lordofhyphens) [inline xs/src/libslic3r/ConfigBase.cpp]: Prefer to use Slic3r::Log

File: xs/src/libslic3r/ConfigBase.cpp [hunk 9]
```diff
@@ -931,7 +936,7 @@ ConfigOptionPoint3::deserialize(std::string str, bool append) {
         this->value.y = boost::lexical_cast<coordf_t>(tokens[1]);
         this->value.z = boost::lexical_cast<coordf_t>(tokens[2]);
     } catch (boost::bad_lexical_cast &e){
-        std::cout << "Exception caught : " << e.what() << std::endl;
+        std::cout << "Config option deserialisation error of ["<<str<<"]  : " << e.what()<<" (expected 3D point)" << std::endl;
```
Comment (lordofhyphens) [inline xs/src/libslic3r/ConfigBase.cpp]: Prefer to use Slic3r::Log, although I would prefer to push more information down into `what()` if at all possible.

File: src/slic3r.cpp [hunk 10]
```diff
@@ -78,22 +76,26 @@ int CLI::run(int argc, char **argv) {
...
-    
+    std::cerr<<"Full_print_config created"<<std::endl;
     // validate config
     try {
         this->full_print_config.validate();
     } catch (InvalidOptionException &e) {
         boost::nowide::cerr << e.what() << std::endl;
         return 1;
     }
+    std::cerr<<"Config validated"<<std::endl;
```
Comment (lordofhyphens) [inline src/slic3r.cpp]: `Slic3r::Log` or remove.

File: src/slic3r.cpp [hunk 11]
```diff
@@ -78,22 +76,26 @@ int CLI::run(int argc, char **argv) {
...
+    }catch (...){
+	std::cerr<<"Exception in 'load' processing "<<std::endl;
+    }
     
     // apply command line options to a more specific DynamicPrintConfig which provides normalize()
     // (command line options override --load files)
     this->print_config.apply(config, true);
     this->print_config.normalize();
-    
+    std::cerr<<"Print_config normalized" << std::endl;
```
Comment (lordofhyphens) [inline src/slic3r.cpp]: `Slic3r::Log` or remove.

File: src/slic3r.cpp [hunk 12]
```diff
@@ -78,22 +76,26 @@ int CLI::run(int argc, char **argv) {
...
     // create a static (full) print config to be used in our logic
     this->full_print_config.apply(this->print_config);
-    
+    std::cerr<<"Full_print_config created"<<std::endl;
```
Comment (lordofhyphens) [inline src/slic3r.cpp]: `Slic3r::Log` or remove.

File: src/slic3r.cpp [hunk 13]
```diff
@@ -78,22 +76,26 @@ int CLI::run(int argc, char **argv) {
         c.normalize();
         this->print_config.apply(c);
     }
+    }catch (...){
```
Comment (lordofhyphens) [inline src/slic3r.cpp]: Not a fan of catch-all, please justify why this is preferable.

File: src/slic3r.cpp [hunk 14]
```diff
@@ -78,22 +76,26 @@ int CLI::run(int argc, char **argv) {
...
+	std::cerr<<"Exception in 'load' processing "<<std::endl;
```
Comment (lordofhyphens) [inline src/slic3r.cpp]: `Slic3r::Log`

File: src/slic3r.cpp [hunk 15]
```diff
@@ -36,27 +36,25 @@ int CLI::run(int argc, char **argv) {
...
-    
     // parse all command line options into a DynamicConfig
     t_config_option_keys opt_order;
     this->config_def.merge(cli_actions_config_def);
     this->config_def.merge(cli_transform_config_def);
     this->config_def.merge(cli_misc_config_def);
     this->config_def.merge(print_config_def);
     this->config.def = &this->config_def;
-    
+    std::cerr<<"Configs merged"<<std::endl;
```
Comment (lordofhyphens) [inline src/slic3r.cpp]: `Slic3r::Log` or remove.

File: src/CMakeLists.txt [hunk 16]
```diff
@@ -11,6 +11,7 @@ option(COVERAGE "Build with gcov code coverage profiling." OFF)
 
 # only on newer GCCs: -ftemplate-backtrace-limit=0
 add_compile_options(-DNO_PERL  -DM_PI=3.14159265358979323846 -D_GLIBCXX_USE_C99 -DHAS_BOOL -DNOGDI -DBOOST_ASIO_DISABLE_KQUEUE)
+#add_link_options(-rdynamic)
```
Comment (lordofhyphens) [inline src/CMakeLists.txt]: Remove commented items in PRs.

File: xs/src/libslic3r/ConfigBase.cpp [hunk 17]
```diff
@@ -394,6 +395,10 @@ ConfigBase::apply_only(const ConfigBase &other, const t_config_option_keys &opt_
             std::string error = "Unexpected failure when deserializing serialized value for " + opt_key;
             CONFESS(error.c_str());
         }
+        } catch ( UnknownOptionException & e ){
+        
+    	    //    std::cerr<<e.what()<<std::endl;
```
Comment (lordofhyphens) [inline xs/src/libslic3r/ConfigBase.cpp]: `Slic3r::Log` or remove. Try not to leave commented code around.

File: xs/src/libslic3r/ConfigBase.cpp [hunk 18]
```diff
@@ -371,6 +371,7 @@ void
 ConfigBase::apply_only(const ConfigBase &other, const t_config_option_keys &opt_keys, bool ignore_nonexistent, bool default_nonexistent) {
     // loop through options and apply them
     for (const t_config_option_key &opt_key : opt_keys) {
+        try{
```
Comment (lordofhyphens) [inline xs/src/libslic3r/ConfigBase.cpp]: Fix containing indentation.

File: src/slic3r.cpp [hunk 19]
```diff
@@ -323,6 +325,7 @@ int CLI::run(int argc, char **argv) {
                 print.center = !this->config.has("center")
                     && !this->config.has("align_xy")
                     && !this->config.getBool("dont_arrange");
+                std::cerr<<"Arrange: "<<print.arrange<<", center: "<<print.center<<std::endl;
```
Comment (lordofhyphens) [inline src/slic3r.cpp]: `Slic3r::Log` or remove.

**Comment by lordofhyphens** [comment]
Appreciate the workaround/fix for no support, but there are numerous issues that need to be addressed. Please fix and update your branch. Thank you.

File: xs/src/libslic3r/PrintGCode.cpp (hunk 4 above)
Comment (luketaverne) [reply xs/src/libslic3r/PrintGCode.cpp]: Will reference this line in my comment on the PR

**Comment by luketaverne** [comment]
Hey @lordofhyphens, thanks for the detailed feedback and for your hard work on Slic3r! @zdroyer and I will work on making the changes in the next day.

Just for clarity: the pull request (after refactoring according to your review), doesn't quite fix the missing support material, it just prevents several segmentation faults when trying to generate support material. The current problem is something like "the layer doesn't have region 0" (I commented on the line I'm talking about above. 

I will write up a more detailed report with specific questions once we apply the requested changes. It would be great to have your insight on where to continue debugging this problem :) 

Thanks again, have a nice evening!

Best,
Luke

**Comment by lordofhyphens** [comment]
@luketaverne yeah, I saw that. While I would much prefer getting supports in, this is a fine effort to make it more functional for a lot of people.

**Comment by lordofhyphens** [comment]
@luketaverne the root cause is that the C++ support code is unfinished.

File: xs/src/libslic3r/ConfigBase.hpp (hunk 3 above)
Comment (zdroyer) [reply xs/src/libslic3r/ConfigBase.hpp]: The override of 'what' had been moved up to the base class do get a generic message of any *OptionException. Anyway, the InvalidValue override was restored in the correction push (69533af)

**Comment by zdroyer** [comment]
@lordofhyphens I followed your suggestions and used Slic3r::Log as much possible. As @luketaverne  mentioned, we're working on fixing the support material generation. The first was to make the commandline not crash and it seems it's done. We're open to suggestions how to improve the support material generation.

**Comment by lordofhyphens** [comment]
Going forwards: Slic3r uses 4 spaces instead of tabs for indentation.

**Comment by lordofhyphens** [comment]
In the future:
* Try not to make more changes than are strictly necessary for the task at hand.
* Slic3r indentation is 4 spaces per indent.

**Comment by lordofhyphens** [comment]
I found a few other issues with the logic when going over the problem myself. I've pushed the changes onto your branch and will merge once the builds finish.

**Comment by zdroyer** [comment]
@lordofhyphens , thanks for the constructive feedback!

**Pull request merged**